proposals/cache_enhancements.md

```diff
@@ -0,0 +1,125 @@
+    kubernetesClientSet, 10*time.Minute, "", func(l0 *metav1.ListOptions) {
+        // setting kind here is optional, if we do not provide it, all resources which have labels will be cached
+        l0.Kind = "ConfigMap"
+        l0.LabelSelector = labelSelector
+    })
+```
+- pass this `kubeResourceInformer` to webhooks handlers on [line](https://github.com/kyverno/kyverno/blob/main/cmd/kyverno/main.go#L617)
+- add `kubeResourceInformer` variable in [handlers struct](https://github.com/kyverno/kyverno/blob/main/pkg/webhooks/resource/handlers.go#L53)
+- initialise `kubeResourceInformer` [here](https://github.com/kyverno/kyverno/blob/main/pkg/webhooks/resource/handlers.go#L88)
+- add kubeResourceInformer in policyContext at following places.
+    - https://github.com/kyverno/kyverno/blob/main/pkg/webhooks/resource/handlers.go#L127
+    - https://github.com/kyverno/kyverno/blob/main/pkg/webhooks/resource/handlers.go#L169 (need to verify if we require to set here as we are calling build again on line 181)
+    - https://github.com/kyverno/kyverno/blob/main/pkg/webhooks/resource/handlers.go#L185
+    - https://github.com/kyverno/kyverno/blob/main/pkg/webhooks/resource/validation/validation.go#L140 (need to verify)
+    NOTE: we are adding `kubeResourceInformer` in policyContext so that in future we won't need to create lister for each resources again.
+- add `KubeResourceInformer` [here](https://github.com/kyverno/kyverno/blob/main/pkg/engine/policyContext.go#L44)
+- get ConfigMap from informer cache [here](https://github.com/kyverno/kyverno/blob/main/pkg/engine/jsonContext.go#L353)
+```bash
+    informerCache := &InformerCache{}
+    obj, err := informerCache.GetK8sResourceFromCache(ctx, "ConfigMap", name.(string), namespace.(string))
+    if err == nil {
+        // use object from informer
+    } else if strings.Contains(err.Error(), "not found") {
+        obj, err := ctx.Client.GetResource("v1", "ConfigMap", namespace.(string), name.(string))
+        if err != nil {
+            return nil, fmt.Errorf("failed to get configmap %s/%s : %v", namespace, name, err)
+        }
+    } else {
+        return nil, fmt.Errorf("failed to get configmap %s/%s : %v", namespace, name, err)
+    }
+```
+create following interface in new file https://github.com/kyverno/kyverno/blob/main/pkg/engine/k8sResourceCache.go
+```bash
+type InformerCache interface {
+    GetK8sResourceFromCache(policyContext *engine.PolicyContext, resourceKind, resourceName, resourceNamespace string) (*unstructured.Unstructured, error)
+}
+
+// Create a function
+func GetK8sResourceFromCache(ctx, kind, name, namespace) (*unstructured.Unstructured, error) {
+    // get the lister for kind using switch case ?
+    // get the object from cache
+    // if object not found, return (nil, err)
+    // if object found, convert it to *unstructure.Unstructured
+    // obj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(resourceObj)
+    // return obj and err
+}
```

Comment:
As discussed on slave, something like this might be a good approach:
```golang
type ConfigmapResolver interface {
  Get(...) (*ConfigMap, err)
}

type InformerBasedResolver struct {
  // holds an informer cache
}

func (r *InformerBasedResolver) Get(...) (*ConfigMap, err) {
  // try to lookup from the cache
}

type ClientBasedResolver struct {
  // holds a client
}

func (r *ClientBasedResolver) Get(...) (*ConfigMap, err) {
  // try to lookup from the client
}

type ResolverChain []ConfigmapResolver

func (r ResolverChain) Get(...) (*ConfigMap, err) {
  for _, resolver := range r {
    if cm, err := resolver.Get(...); err == nil {
      return cm, nil
    }
  }
  return nil, errors.New("not found")
}
```

This needs to be easily adapted depending on the context, the CLI will not use the same features as the kyverno admission controller for example.

Reply:
added this